### ISSUE TITLE
“meteor” terminated by signal SIGSEGV (Address boundary error)

### ISSUE THREAD

**Issue by yorrd** [meteor/meteor#9705]
Meteor is crashing on startup. This is happening on several of my projects, though I can't figure out a single reason to pin this on. Most importantly: it started to happen more or less randomly a few weeks back. 

- As far as I can tell it's not connected to the meteor version (tested with 1.6.* and 1.5.4). 
- There have been no changes to the projects except a few lines of JavaScript in the client side code since before this was happening - it just didn't work anymore one day. 
- A colleague of mine who works on the same project does NOT have the issue (we're working this on a git repository) which makes me think this is some kind of discrepancy on our systems. However, we're basically running on the same linux distribution (arch), pretty much always up to date, so this would have to be something very specific (EDIT: my colleague has the same precise packages and node versions, so we're precisely using the same project files)
- `meteor reset` does not help at all.
- it fails at "building for web.browser"
- can't reproduce with a fresh project

This is the (pretty non-helpful) log:

```
[[[[[ ~/PhpstormProjects/projectname ]]]]]

=> Started proxy.
fish: “meteor” terminated by signal SIGSEGV (Address boundary error)
```

This is the `lldb bt`:

```
  * frame #0: 0x0000000000ee7501 node`v8::internal::Object* v8::internal::VisitWeakList<v8::internal::JSFunction>(v8::internal::Heap*, v8::internal::Object*, v8::internal::WeakObjectRetainer*) + 177
    frame #1: 0x0000000000ee7c80 node`v8::internal::Object* v8::internal::VisitWeakList<v8::internal::Context>(v8::internal::Heap*, v8::internal::Object*, v8::internal::WeakObjectRetainer*) + 240
    frame #2: 0x0000000000ea5817 node`v8::internal::Heap::Scavenge() + 5031
    frame #3: 0x0000000000ea6d24 node`v8::internal::Heap::PerformGarbageCollection(v8::internal::GarbageCollector, v8::GCCallbackFlags) + 980
    frame #4: 0x0000000000ea79e1 node`v8::internal::Heap::CollectGarbage(v8::internal::AllocationSpace, v8::internal::GarbageCollectionReason, v8::GCCallbackFlags) + 929
    frame #5: 0x0000000000e4b192 node`v8::internal::Factory::NewFillerObject(int, bool, v8::internal::AllocationSpace) + 82
    frame #6: 0x00000000010c8810 node`v8::internal::Runtime_AllocateInNewSpace(int, v8::internal::Object**, v8::internal::Isolate*) + 96
    frame #7: 0x000036fde4d0463d
    frame #8: 0x000036fde5112857
    frame #9: 0x000036fde4d05d1b
    frame #10: 0x000036fde4df461c
    frame #11: 0x000036fde4dbf9ce
    frame #12: 0x000036fde50248d1
    frame #13: 0x000036fde4d05d1b
    frame #14: 0x000036fde4fd57fe
    ... more empty frames ...
    frame #174: 0x000036fde4df461c
    frame #175: 0x000036fde4dbf9ce
    frame #176: 0x000036fde4df44e9
    frame #177: 0x000036fde4dbf9ce
    frame #178: 0x000036fde4dbe479
    frame #179: 0x000036fde4d0410d
    frame #180: 0x0000000000e40c6a node`v8::internal::Execution::Call(v8::internal::Isolate*, v8::internal::Handle<v8::internal::Object>, v8::internal::Handle<v8::internal::Object>, int, v8::internal::Handle<v8::internal::Object>*) + 266
    frame #181: 0x0000000000afdc23 node`v8::Function::Call(v8::Local<v8::Context>, v8::Local<v8::Value>, int, v8::Local<v8::Value>*) + 355
    frame #182: 0x0000000000b0c8c1 node`v8::Function::Call(v8::Local<v8::Value>, int, v8::Local<v8::Value>*) + 65
    frame #183: 0x00007f61d629604f fibers.node`Fiber::RunFiber(void**) + 831
    frame #184: 0x00007f61d6296b8d fibers.node`Coroutine::trampoline(void*) + 205
    frame #185: 0x00007f61d6296fb7 fibers.node`coro_init + 55
    frame #186: 0x00007f61d84e9080 libc.so.6
```

I've seen several issues on this dating a few years back but no real solution except for the issue being closed because it should be fixed. 

Please let me know if I can provide any further information. I'm not sure how to

**Comment by hwillson** [comment]
Have you tried reinstalling meteor via `curl https://install.meteor.com/ | sh`? Since this appears to be isolated to your system, reinstalling might help.

**Comment by yorrd** [comment]
Thanks for the suggestion, did not change. I had a mongo exit code 14, but solved it by resetting the project. Now the behavior is the same.

**Comment by yorrd** [comment]
further information:

- it fails at "building for web.browser"
- can't reproduce with a fresh project

does the backtrace help at all? If it doesn't, I guess I'll have to find something else to work off of.

**Comment by hwillson** [comment]
The backtrace definitely points towards this being a memory issue. Maybe try adjusting the memory available to v8 using something like:

```
TOOL_NODE_FLAGS="--max-old-space-size=4096" meteor ...
```

(adjust the size as needed)

**Comment by axelvaindal** [comment]
@yorrd 

Some other people was having similar issues with some `npm` package (gulp for example) due to legacy files on their project from various updates. 
Have you tried removing `node_modules` folder and run `meteor npm install` ? 
Are you sure every package are compatible with the version of `nodejs` you are using on your system ?
Based on your say, isn't there anything which could consume the available memory in a way `meteor run` is failing ? (As @hwillson pointed out with a workaround).

**Comment by yorrd** [comment]
@hwillson 
@axelvaindal 

I've tried that flag, I've reinstalled all node modules. No change. 
Also, I've been watching my memory usage. The entirely free RAM (without cached or anything) never falls below over 1GB. That can hardly be it then, can it?

Thanks for the help

**Comment by hwillson** [comment]
Wow, definitely a strange one @yorrd. You mentioned new apps work - maybe try creating a new empty application, then copy pieces of your broken app over bit by bit, to see if you can either get it working, or pinpoint what's causing it to fail?

**Comment by yorrd** [comment]
I have new information.

For some reason, this now happens randomly instead of the sigsev. By randomly I mean: sometimes this error comes and sometimes the other, when starting meteor the exact same way 100 times in a row. 

Additional information: because of https://github.com/meteor/meteor-feature-requests/issues/6, I'm copying the complete Polymer library and many Polymer elements into a directory within `/imports`. When I don't do that, the error doesn't happen. I'm assuming it just can't handle that many files. When removing slowly from that folder, it doesn't matter what I remove, at some point it just works again when there are fewer libraries in there. 

I'm wondering though: Why does it not complain with my colleague?

Is there anything I can do? I need those libraries and because of the frequent updates in prerelease phase, I don't feel like I want to babel compile them all by hand each time an update is there...

```
=> Errors prevented startup:                  
   
   While processing files with static-html (for target web.browser):

   /home/yorrd/.meteor/packages/static-html/.1.2.2.1pf71mk.8tol++os+web.browser+web.cordova/plugin.compileStaticHtmlBatch.os/npm/node_modules/meteor/caching-compiler/node_modules/async/lib/async.js:320:29:
   Maximum call stack size exceeded
   at
   /home/yorrd/.meteor/packages/static-html/.1.2.2.1pf71mk.8tol++os+web.browser+web.cordova/plugin.compileStaticHtmlBatch.os/npm/node_modules/meteor/caching-compiler/node_modules/async/lib/async.js:320:29

... thousands more of these ...

   at
   /home/yorrd/.meteor/packages/static-html/.1.2.2.1pf71mk.8tol++os+web.browser+web.cordova/plugin.compileStaticHtmlBatch.os/npm/node_modules/meteor/caching-compiler/node_modules/async/lib/async.js:44:16
   at packages/caching-compiler/caching-compiler.js:316:9
   at
   /home/yorrd/.meteor/packages/static-html/.1.2.2.1pf71mk.8tol++os+web.browser+web.cordova/plugin.compileStaticHtmlBatch.os/npm/node_modules/meteor/caching-compiler/node_modules/async/lib/async.js:182:20
   at replenish
   (/home/yorrd/.meteor/packages/static-html/.1.2.2.1pf71mk.8tol++os+web.browser+web.cordova/plugin.compileStaticHtmlBatch.os/npm/node_modules/meteor/caching-compiler/node_modules/async/lib/async.js:317:21)
   at
   /home/yorrd/.meteor/packages/static-html/.1.2.2.1pf71mk.8tol++os+web.browser+web.cordova/plugin.compileStaticHtmlBatch.os/npm/node_modules/meteor/caching-compiler/node_modules/async/lib/async.js:324:29
   at Object.async.forEachLimit.async.eachLimit
   (/home/yorrd/.meteor/packages/static-html/.1.2.2.1pf71mk.8tol++os+web.browser+web.cordova/plugin.compileStaticHtmlBatch.os/npm/node_modules/meteor/caching-compiler/node_modules/async/lib/async.js:221:35)
   at CachingHtmlCompiler.processFilesForTarget
   (packages/caching-compiler/caching-compiler.js:284:11)
   at CachingHtmlCompiler.processFilesForTarget
   (packages/caching-html-compiler/caching-html-compiler.js:40:5)
```

**Comment by hwillson** [comment]
Can you try running meteor with file polling enabled, to see if that helps?

```
METEOR_WATCH_FORCE_POLLING=1 meteor ...
```

**Comment by yorrd** [comment]
I'm really really sorry, but I can't reproduce it anymore. It magically disappeared. Something must have changed on my system. I'll reopen this and continue testing when the issue reappears.

**Comment by yorrd** [comment]
HAH! I got it again. Different project now, the other one doesn't complain anymore. The funny thing: I didn't change anything on the old project. The new project has been written entirely from scratch since then. 

More information:

I have a LOT of libraries copied over into the /imports directory, because they need to be compiled (and meteor currently still ignores the node_modules directory, or I have missed the change). If I take out half of them, it reliably works again. Something else I have noticed: when running in `--production` mode, I almost always get the `SIGSEGV`, when running in dev mode, I almost always get the `caching-compiler` error.

Anything else I can provide you with?

**Comment by yorrd** [comment]
Maybe also worth noting: in none of the new beta versions does anything change. Is there reason to believe an old version could work better? Which one would you want me to test?

**Comment by axelvaindal** [comment]
Hi @yorrd,

This seems deeply related to processing the compilation of your files from the `imports` directory on startup. I'm not really familiar with this (as most of my `npm` related libraries come with a compiled version), but could you tell us the amount of libraries we are talking about here ?

In the FR you linked above, @benjamn explains the main reason `npm` module are not compiled at the moment is a matter of performance. Based on the errors you're facing, I suppose the system you're running meteor isn't able to cache that much files and by attempting to do so is trying to use reserved memory.
What is your system configuration ? Are you sure your hardware is efficient enough to process a project this big (based on your statement) ?
I have myself faced some notable speed improvement on building to web browser moving from Linux 14.04 (Old Hardware) to Linux 16.04 (New Hardware) on my development environment, so I guess it most likely the issue here.

Then again, I'm not an expert on how Meteor building works in the background (altough I would be deeply interested in learning and being able to contribute more), nor am I on caching/compiling from ES6, so maybe @hwillson will be able to tell you more 🙂

**Comment by yorrd** [comment]
@axelvaindal thanks for your reply, let me try to answer everything adequately.

I'm aware that compiling all dependencies is a huge performance-issue. However, at the moment I have no choice but compile them myself which is probably even worse and I'll have to restart every time I add a dependency (which is particularly annoying because I'm all the time adding polymer components).

My system is running on 16GB of RAM which was at the time of testing not even half-way used, including cached usage. My Linux is Arch, completely up to date.

@everyoneInterested: I now have a start script which compiles with babel. This is outside of meteor though and will require a restart. I'm not sure if a compiler-plugin for meteor is allowed to compile `node_modules` contents. If yes, please point me in the right direction here and I'll happily write one.

**Comment by axelvaindal** [comment]
Hum.
If I understand you correctly (I'm not a native english speaker, so I may misunderstand things 😄), you're using Arch Linux latest version, with 16 Gb of RAM not entirely consumed during the Meteor internal build process and yet, still get the segmentation fault or the cache-compiler error.
So, you use `babel` to compile the `Polymer` library and custom components you're using, and copy them in the `imports` directory from `node_modules`. Then, the script runs Meteor as usual.
With that procedure, everything works fine, though you have to rebuild "manually" every time you add a new component (thus, restarting Meteor as well), which is happening quite often and thus is not ideal.
Is that correct ?

If so, there is something definitely weird happening here. I don't understand why the `babel` compiler used manually would work over the one shipped with Meteor. If you may provide a reproduction I could clone to test it on my hardware, I could try to build on my own and see if the error is happening.
What I am interested in here is : 

- Why is the manual build working over the automated one ? How long to build manually ? Does it throw any kind of errors or warning ? What's the memory usage on both automatic and manual building ?
- Is it related to specific files or not ?
- What happen if we use the same amount of files to process with blank component (no content, no behavior, just a `class MyPolymerElement extends Polymer.Element {...}` ) ? 
- How does all this behave on a system which have less RAM available ?

For your request about compiling `node_modules`, as @benjamn stated it is a top priority for Meteor in the v1.6.2, I assume they are working on it and it will be handled without the need of a custom compiler-plugin. I understand however that you might be in hurry, but I'm pretty confident in saying that fixing the strange behavior of Meteor with your `imports` folder will most likely be a better long-term solution to your problem. It would allow you to keep working on your app without having to develop further tools which are included in Meteor by default. Moreover, when the `node_modules` transpilation comes out, you will be all set and you will just have to use the `node_modules` directly 🙂.

**Comment by yorrd** [comment]
You understand entirely correctly. I'm not native either, so forgive me if I'm confusing you in any way :D

Here's the complete truth: I copy all polymer components, the polymer library itself and the webcomponentsjs library to imports/ui/node_links. Then I rename all .js files to .ts. Yes, that's dirty, but it works. That makes the typescript meteor plugin compile the files to es5 which is exactly what I want. At that point, those libraries might just as well be my code as far as Meteor is concerned. Unfortunately, it's a LOT of code, so I understand the memory issues. And of course, I have an alternative solution now, which just compiles the libraries in place in the node_modules directory, but that's suboptimal because 1) I don't want to change to node_modules directory and 2) I have to restart when adding new libraries (which, as I said before, happens a LOT). 

The babel compiler I'm using is just run once before Meteor is run and it is only run on each single .js file. That's why it doesn't get those issues I assume. And it takes a long time as well, but that's half as bad. To summarize: I don't think you can compare the two solutions because one runs within Meteor and one runs only once and before Meteor even starts.

No, no specific files, it doesn't matter which half I take out, when there's less stuff, it works, plainly speaking. 

I'm not sure what would happen if all the libraries would only contain empty classes, but they would have to be loaded because otherwise Meteor won't bundle them (imports only gets loaded when something imports it as far as I know).

I have already restrained my RAM artificially to 4GB which didn't change anything.

Love hearing that node_modules compilation is a priority, didn't see official statements on that yet :)

I do have a workaround at the moment though (precompiling within node_modules in the npm start script), so this is not a pressing issue for me at the moment. 

@hwillson I would understand if you said that this is an issue which nobody should have because nobody should ever compile so much stuff. I pretty much only reopened this because I thought it might help you find a potential issue. As stated, I do have a workaround and as soon as node_modules compilation is done, this will be a non-issue anyways, because nobody will ever run projects with a code base this large not including libraries...

tl;dr: I'm happy to assist in finding the underlying issue here or provide a reproduction which fails on my system, but feel free to close if this is not something you're looking to fixing anytime soon.
